common/simple_dom.h

```diff
@@ -0,0 +1,274 @@
+        if (_node)
+            _node->root()->del_ref();
+        _node = rhs._node;
+        rhs._node = nullptr;
+        return *this;
+    }
+    ~NodeWrapper() {
+        _node->root()->del_ref();
+    }
+    NodeWrapper root() const       { IF_RET(_node->root()); }
+    RootNode* root_node() const    { IF_RET(_node->root()); }
+    NodeWrapper parent() const     { IF_RET(_node->_parent); }
+    NodeWrapper next() const       { IF_RET(_node->_next); }
+    rstring_view32 rkey() const    { IF_RET(_node->_key); }
+    rstring_view32 rvalue() const  { IF_RET(_node->_value); }
+    str key(const char* b) const   { IF_RET(b | rkey()); }
+    str value(const char* b) const { IF_RET(b | rvalue()); }
+    const char* text_begin() const {
+        IF_RET(root()._node->_text_begin);
+    }
+    str key() const    { IF_RET(text_begin() | rkey()); }
+    str value() const  { IF_RET(text_begin() | rvalue()); }
+    bool valid() const { return _node; }
+    operator bool() const { return _node; }
+
+    size_t num_children() const {
+        return _node ? _node->num_children() : 0;
+    }
+    NodeWrapper get(size_t i) const {
+        IF_RET({_node->get(i)});
+    }
+    NodeWrapper get(str key) const {
+        IF_RET({_node->get(key)});
+    }
+    template<size_t N>
+    NodeWrapper operator[](const char (&key)[N]) const {
+        return get(key);
+    }
+    NodeWrapper operator[](str key) const {
+        return get(key);
+    }
+    NodeWrapper operator[](size_t i) const {
+        return get(i);
+    }
+    NodeWrapper get_attributes() const {
+        return get("__attributes__");
+    }
+    str to_string() const {
+        return value();
+    }
+    int64_t to_integer(int64_t def_val = 0) const {
+        return value().to_uint64(def_val);
+    }
+    double to_number(double def_val = NAN) const {
+        return value().to_double(def_val);
+    }
+
+    struct SameKeyEnumerator;
+    Enumerable<SameKeyEnumerator> enumerable_same_key_siblings() const;
+
+    struct ChildrenEnumerator;
+    Enumerable<ChildrenEnumerator> enumerable_children() const;
+
+    Enumerable<SameKeyEnumerator>  enumerable_children(str key) const {
+        return get(key).enumerable_same_key_siblings();
+    }
+};
+#undef IF_RET
+
+// lower 8-bit are reserved for doc types
+const int DOC_JSON = 0x00;
+const int DOC_XML  = 0x01;
+const int DOC_YAML = 0x02;
+const int DOC_INI  = 0x03;
+const int DOC_TYPE_MASK = 0xff;
+
+const int FLAG_FREE_TEXT_IF_PARSING_FAILED = 0x100;
+
+using Document = NodeWrapper;
+
+inline NodeWrapper Node::wrapper() { return {this}; }
+
+// 1. text is moved to the simple_dom object, which frees it when destruct.
+// 2. the content of text may be modified in-place to un-escape strings.
+// 3. returning a pointer (of Node) is more efficient than an object (of Document),
+//    even if they are equivalent in binary form.
+Node* parse(char* text, size_t size, int flags);
+
+inline Node* parse(IStream::ReadAll&& buf, int flags) {
```

Comment:
Isn't the Node meant for internal use? How come it's being used by users here?

Reply:
parse(...)->wrap()
```
// 3. returning a pointer (of Node) is more efficient than an object (of Document),
//    even if they are equivalent in binary form.
```

Reply:
For users, it's a direct use after parsing, and calling wrap is a burden.

Reply:
Not a big problem

Reply:
It a problem. The returned NodeImpl* is not the type used by the user and users must first determine if the returned NodeImpl* is nullptr before calling wrap. it's best not to expose these to users.

Reply:
OK